File: Drakborgen/Board.cpp
#include "Board.h"

#include <iostream>
#include <vector>

#include "System.h"
#include "Tile.h"
#include "Tower.h"
#include "Vault.h"
#include "Setup.h"

Board::Board(AnimationManager& animations)
	: clickOverlay(sf::Vector2f(tileSize, tileSize))
	, boardSprite(Setup::getMediaPath() + "spelplan.jpg")
	, vaultSprite(Setup::getMediaPath() + "skattkammaren.png")
	, animations(animations)
{
	vaultSprite.get().setPosition(getSitePosition({ 4, 6 }));
	vaultSprite.get().move(sf::Vector2f(0.0f, 10.0f));
	//players.reserve(4);

	tileGrid[0][0] = std::make_unique<Tower>(Direction::East, Direction::South);
	tileGrid[0][columnCount - 1] = std::make_unique<Tower>(Direction::West, Direction::South);
	tileGrid[rowCount - 1][0] = std::make_unique<Tower>(Direction::East, Direction::North);
	tileGrid[rowCount - 1][columnCount - 1] = std::make_unique<Tower>(Direction::West, Direction::North);
	tileGrid[4][6] = Vault::makeVault(Direction::North);
	tileGrid[5][6] = Vault::makeVault(Direction::South);
}

void Board::update(float elapsedTime, float /*timeDelta*/)
{
	// one cycle per second
	if (moveSiteAnimationStartTime == 0.0f)
	{
		moveSiteAnimationStartTime = elapsedTime;
	}
	const int wholeSeconds = int(elapsedTime - moveSiteAnimationStartTime);
	float fraction = elapsedTime - moveSiteAnimationStartTime - wholeSeconds;
	if (wholeSeconds % 2 != 0)
	{
		fraction = 1.0f - fraction;
	}
	clickOverlay.setFillColor(sf::Color(uint8_t(0x80 + 0x7f * fraction), uint8_t(0x80 + 0x7f * fraction), uint8_t(0x80 + 0x7f * fraction), uint8_t(0xff * fraction)));
}

void Board::placeTile(std::unique_ptr<Tile> tile, MoveSite moveSite)
{
	if (hasTile(moveSite.site) || !withinBounds(moveSite.site))
	{
		THROW;
	}
	tile->setPosition(getSitePosition(moveSite.site));
	tile->setOrientation(moveSite.direction);
	tileGrid[moveSite.site.row][moveSite.site.column] = std::move(tile);
}

sf::Vector2f Board::getSitePosition(Site site) const
{
	return sf::Vector2f(gridOriginX + tileSize * site.column, gridOriginY + tileSize * site.row);
}

Board::Site Board::getSite(sf::Vector2f position) const
{
	return { int((position.y - gridOriginY) / tileSize), int((position.x - gridOriginX) / tileSize) };
}

sf::Vector2f Board::getAvatarCenter(int index) const
{
	sf::FloatRect bounds = players[index].avatar->getGlobalBounds();
	return { bounds.left + bounds.width / 2.0f, bounds.top + bounds.height / 2.0f };
}

sf::Vector2f Board::getSize() const
{
	return boardSprite.get().getLocalBounds().getSize();
}

void Board::setGameStartMoveSites()
{
	moveSites = { { 0, 0 }, { 0, columnCount - 1 }, { rowCount - 1, 0 }, { rowCount - 1, columnCount - 1 } };
	moveSiteAnimationStartTime = 0.0f;
}

void Board::setPlayerMoveSites(MoveSite playerSite)
{
	for (const std::vector<Direction>& exits : getTile(playerSite.site)->getExits(playerSite.direction))
	{
		MoveSite moveSite = createMoveSite(playerSite.site, exits);
		if (withinBounds(moveSite.site))
		{
			moveSites.push_back(moveSite);
		}
	}
	moveSiteAnimationStartTime = 0.0f;
}

bool Board::testMoveSites(sf::Vector2f position) const
{
	if (!moveSitesShown)
	{
		THROW;
	}
	Site site = getSite(position);
	for (MoveSite moveSite : moveSites)
	{
		if (site == moveSite.site)
		{
			return true;
		}
	}
	return false;
}

Board::MoveSite Board::getMoveSite(sf::Vector2f position) const
{
	if (!moveSitesShown)
	{
		THROW;
	}
	Site site = getSite(position);
	for (MoveSite moveSite : moveSites)
	{
		if (site == moveSite.site)
		{
			return moveSite;
		}
	}
	THROW;
}

void Board::removeMoveSite(Site site)
{
	for (std::vector<MoveSite>::iterator it = moveSites.begin(); it < moveSites.end(); ++it)
	{
		if (it->site == site)
		{
			moveSites.erase(it);
			break;
		}
	}
}

void Board::clearMoveSites()
{
	moveSites.clear();
}

void Board::showMoveSites(bool show)
{
	moveSitesShown = show;
	if (show)
	{
		moveSiteAnimationStartTime = 0.0f;
	}
}

void Board::addPlayer(const std::string& imagePath, int index)
{
	if (players.size() != index)
	{
		THROW;
	}
	players.push_back({ std::make_unique<Card>(imagePath), invalidSite });
	Player& player = players.back();
	player.avatar->setScale(sf::Vector2f(1.0f / 3.0f, 1.0f / 3.0f));
}

void Board::setPlayerSite(int index, MoveSite moveSite, std::function<void()> callback)
{
	if (players.size() <= index)
	{
		THROW;
	}
	players[index].site = moveSite.site;
	placePlayer(index, moveSite.direction, callback);
}

void Board::draw(sf::RenderTarget& target, sf::RenderStates states) const
{
	states.transform *= getTransform();
	target.draw(boardSprite.get(), states);
	target.draw(vaultSprite.get(), states);
	for (const std::array<std::unique_ptr<Tile>, 13>& row : tileGrid)
	{
		for (const std::unique_ptr<Tile>& tile : row)
		{
			if (tile)
			{
				target.draw(*tile, states);
			}
		}
	}
	if (moveSitesShown)
	{
		for(MoveSite moveSite : moveSites)
		{
			sf::Transform transform = states.transform;
			sf::Vector2f offset = getSitePosition(moveSite.site);
			states.transform.translate(offset);
			target.draw(clickOverlay, states);
			states.transform = transform;
		}
	}
	for (const Player& player : players)
	{
		target.draw(*player.avatar, states);
	}
}

bool Board::hasTile(Site site) const
{
	std::cout << site.row << ":" << site.column << std::endl;
	return tileGrid[site.row][site.column] != nullptr;
}

std::unique_ptr<Tile>& Board::getTile(Site site)
{
	if (!hasTile(site))
	{
		THROW;
	}
	return tileGrid[site.row][site.column];
}

bool Board::withinBounds(Site site) const
{
	return site.row >= 0 && site.row < rowCount && site.column >= 0 && site.column < columnCount;
}

void Board::placePlayer(int index, Direction direction, std::function<void()> callback)
{
	Player& player = players[index];
	sf::Vector2f sitePosition = getSitePosition(player.site);
	sf::Vector2f avatarSize = player.avatar->getGlobalBounds().getSize();
	sf::Vector2f animationTarget(sitePosition.x + tileSize / 2.0f - avatarSize.x / 2.0f, sitePosition.y + tileSize / 2.0f - avatarSize.y / 2.0f);
	if (direction == Direction::Invalid)
	{
		float avatarStartX = animationTarget.x + (player.site.column == 0 ? -tileSize : tileSize) * 2.0f;
		float avatarStartY = animationTarget.y + (player.site.row == 0 ? -tileSize : tileSize);
		player.avatar->setPosition(sf::Vector2f(avatarStartX, avatarStartY));
	}
	animations.add(*player.avatar, animationTarget, 0.5f, callback);
}

Board::MoveSite Board::createMoveSite(Site site, const std::vector<Direction>& exits)
{
	for (Direction exit : exits)
	{
		switch (exit)
		{
		case Direction::North:
			--site.row;
			break;
		case Direction::South:
			++site.row;
			break;
		case Direction::West:
			--site.column;
			break;
		case Direction::East:
			++site.column;
			break;
		default:
			THROW;
			break;
		}
	}
	return { site, exits.back() };
}

File: DrakborgenDotNet/Board.cs
using SFML.Graphics;
using SFML.System;
using System.Numerics;

namespace Drakborgen
{
    internal class Board : Transformable, Drawable
    {
        internal Vector2f Size
        {
            get
            {
                return _boardSprite.GetLocalBounds().Size;
            }
        }

        private const float GridOriginX = 418.0f;
        private const float GridOriginY = 119.0f;
        private const float TileSize = 178.0f;

        internal readonly record struct Site(int Row, int Column);

        internal readonly record struct MoveSite(Site Site, Direction Direction = Direction.Invalid)
        {
	internal MoveSite(int row, int column, Direction direction = Direction.Invalid)
	    : this(new Site(row, column), direction)
	{
	}
        }

        private record Player(Card Avatar)
        {
	internal Site Site { get; set; } = InvalidSite;
        }

        internal static readonly Site InvalidSite = new Site(-1, -1);
        internal static readonly MoveSite InvalidMoveSite = new MoveSite(InvalidSite);

        internal const int RowCount = 10;
        internal const int ColumnCount = 13;

        private Sprite _boardSprite = new Sprite(new Texture(Setup.MediaPath + "spelplan.jpg"));
        private Sprite _vaultSprite = new Sprite(new Texture(Setup.MediaPath + "skattkammaren.png"));
        private RectangleShape _clickOverlay = new RectangleShape(new Vector2f(TileSize, TileSize));
        private List<Player> _players = new List<Player>(4);

        private Tile[,] _tileGrid = new Tile[RowCount, ColumnCount];
        private List<MoveSite> _moveSites = new List<MoveSite>();
        private bool _moveSitesShown = false;
        private float _moveSiteAnimationStartTime = 0.0f;

        private AnimationManager _animations;

        internal Board(AnimationManager animations)
        {
	_animations = animations;
	_vaultSprite.Position = GetSitePosition(new Site(4, 6));
	_vaultSprite.Position += new Vector2f(0.0f, 10.0f);
	//players.reserve(4);

	_tileGrid[0, 0] = new Tower(new List<Direction>() { Direction.East, Direction.South });
	_tileGrid[0, ColumnCount - 1] = new Tower(new List<Direction>() { Direction.West, Direction.South });
	_tileGrid[RowCount - 1, 0] = new Tower(new List<Direction>() { Direction.East, Direction.North });
	_tileGrid[RowCount - 1, ColumnCount - 1] = new Tower(new List<Direction>() { Direction.West, Direction.North });
	_tileGrid[4, 6] = Vault.MakeVault(Direction.North);
	_tileGrid[5, 6] = Vault.MakeVault(Direction.South);
        }



        internal void Update(float elapsedTime, float timeDelta)
        {
	// one cycle per second
	if (_moveSiteAnimationStartTime == 0.0f)
	{
	    _moveSiteAnimationStartTime = elapsedTime;
	}
	int wholeSeconds = (int)(elapsedTime - _moveSiteAnimationStartTime);
	float fraction = elapsedTime - _moveSiteAnimationStartTime - wholeSeconds;
	if (wholeSeconds % 2 != 0)
	{
	    fraction = 1.0f - fraction;
	}
	_clickOverlay.FillColor = new Color((byte)(0x80 + 0x7f * fraction), (byte)(0x80 + 0x7f * fraction), (byte)(0x80 + 0x7f * fraction), (byte)(0xff * fraction));
        }

        private void PlaceTile(Tile tile, MoveSite moveSite)
        {
	if (HasTile(moveSite.Site) || !WithinBounds(moveSite.Site))
	{
	    throw new Exception();
	}
	tile.Position = GetSitePosition(moveSite.Site);
	tile.SetOrientation(moveSite.Direction);
	_tileGrid[moveSite.Site.Row, moveSite.Site.Column] = tile;
        }

        private Vector2f GetSitePosition(Site site)
        {
	return new Vector2f(GridOriginX + TileSize * site.Column, GridOriginY + TileSize * site.Row);
        }

        internal Site GetSite(Vector2f position)
        {
	return new Site((int)((position.Y - GridOriginY) / TileSize), (int)((position.X - GridOriginX) / TileSize));
        }

        internal Vector2f GetAvatarCenter(int index)
        {
	FloatRect bounds = _players[index].Avatar.GetGlobalBounds();
	return new Vector2f(bounds.Left + bounds.Width / 2.0f, bounds.Top + bounds.Height / 2.0f);
        }

        internal void SetGameStartMoveSites()
        {
	_moveSites = new List<MoveSite>() { new MoveSite(0, 0), new MoveSite(0, ColumnCount - 1), new MoveSite(RowCount - 1, 0), new MoveSite(RowCount - 1, ColumnCount - 1) };
	_moveSiteAnimationStartTime = 0.0f;
        }

        internal void SetPlayerMoveSites(MoveSite playerSite)
        {
	foreach (List<Direction> exits in GetTile(playerSite.Site).GetExits(playerSite.Direction))
	{
	    MoveSite moveSite = CreateMoveSite(playerSite.Site, exits);
	    if (WithinBounds(moveSite.Site))
	    {
	        _moveSites.Add(moveSite);
	    }
	}
	_moveSiteAnimationStartTime = 0.0f;
        }

        internal bool TestMoveSites(Vector2f position)
        {
	if (!_moveSitesShown)
	{
	    throw new Exception();
	}
	Site site = GetSite(position);
	foreach (MoveSite moveSite in _moveSites)
	{
	    if (site == moveSite.Site)
	    {
	        return true;
	    }
	}
	return false;
        }

        internal MoveSite GetMoveSite(Vector2f position)
        {
	if (!_moveSitesShown)
	{
	    throw new Exception();
	}
	Site site = GetSite(position);
	foreach (MoveSite moveSite in _moveSites)
	{
	    if (site == moveSite.Site)
	    {
	        return moveSite;
	    }
	}
	throw new Exception();
        }

        internal void RemoveMoveSite(Site site)
        {
	for (int i = 0; i < _moveSites.Count; i++)
	{
	    if (_moveSites[i].Site == site)
	    {
	        _moveSites.RemoveAt(i);
	        break;
	    }
	}
        }

        internal void ClearMoveSites()
        {
	_moveSites.Clear();
        }

        internal void ShowMoveSites(bool show)
        {
	_moveSitesShown = show;
	if (show)
	{
	    _moveSiteAnimationStartTime = 0.0f;
	}
        }

        internal void AddPlayer(string imagePath, int index)
        {
	if (_players.Count != index)
	{
	    throw new Exception();
	}
	_players.Add(new Player(new Card(imagePath)));
	Player player = _players.Last();
	player.Avatar.Scale = new Vector2f(1.0f / 3.0f, 1.0f / 3.0f);
        }

        internal void SetPlayerSite(int index, MoveSite moveSite, Action animationCallback)
        {
	if (_players.Count <= index)
	{
	    throw new Exception();
	}
	_players[index].Site = moveSite.Site;
	PlacePlayer(index, moveSite.Direction, animationCallback);
        }

        void Drawable.Draw(RenderTarget target, RenderStates states)
        {
	states.Transform *= Transform;
	target.Draw(_boardSprite, states);
	target.Draw(_vaultSprite, states);
	foreach (Tile tile in _tileGrid)
	{
	    if (tile.CanDraw())
	    {
	        target.Draw(tile, states);
	    }
	}
	if (_moveSitesShown)
	{
	    foreach (MoveSite moveSite in _moveSites)
	    {
	        Transform transform = states.Transform;
	        Vector2f offset = GetSitePosition(moveSite.Site);
	        states.Transform.Translate(offset);
	        target.Draw(_clickOverlay, states);
	        states.Transform = transform;
	    }
	}
	foreach (Player player in _players)
	{
	    target.Draw(player.Avatar, states);
	}
        }

        internal bool HasTile(Site site)
        {
	return _tileGrid[site.Row, site.Column] != null;
        }

        internal Tile GetTile(Site site)
        {
	if (!HasTile(site))
	{
	    throw new Exception();
	}
	return _tileGrid[site.Row, site.Column];
        }

        private bool WithinBounds(Site site)
        {
	return site.Row >= 0 && site.Row < RowCount && site.Column >= 0 && site.Column < ColumnCount;
        }

        private void PlacePlayer(int index, Direction direction, Action animationCallback)
{
	Player player = _players[index];
	Vector2f sitePosition = GetSitePosition(player.Site);
	Vector2f avatarSize = player.Avatar.GetGlobalBounds().Size;
	Vector2f animationTarget = new Vector2f(sitePosition.X + TileSize / 2.0f - avatarSize.X / 2.0f, sitePosition.Y + TileSize / 2.0f - avatarSize.Y / 2.0f);
	if (direction == Direction.Invalid)
	{
	    float avatarStartX = animationTarget.X + (player.Site.Column == 0 ? -TileSize : TileSize) * 2.0f;
	    float avatarStartY = animationTarget.Y + (player.Site.Row == 0 ? -TileSize : TileSize);
	    player.Avatar.Position = new Vector2f(avatarStartX, avatarStartY);
	}
	_animations.Add(player.Avatar, animationTarget, 0.5f, animationCallback);
}

        private MoveSite CreateMoveSite(Site site, List<Direction> exits)
        {
	int row = site.Row;
	int column = site.Column;

	foreach (Direction exit in exits)
	{
	    switch (exit)
	    {
	        case Direction.North:
		--row;
		break;
	        case Direction.South:
		++row;
		break;
	        case Direction.West:
		--column;
		break;
	        case Direction.East:
		++column;
		break;
	        default:
		throw new Exception();
	    }
	}

	return new MoveSite(row, column, exits.Last());
        }
    }
}